File: 234-palindrome-linked-list/palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head){
        ListNode *cur=head,*prev=NULL;

        while(cur){
            ListNode* temp=cur->next;
          cur->next=prev;
          prev=cur;
          cur=temp;
        }
        // cout<<prev->val;

        return prev;
    }
    bool isPalindrome(ListNode* head) {
         ListNode *fast=head,*slow=head;
     while(fast!=NULL && fast->next!=NULL){
        slow=slow->next;
        fast=fast->next->next;
        }

        ListNode *cur;
       if(fast) cur=slow->next;
       else cur=slow;

     fast=reverseList(cur);
        
        while(head!=slow && fast){
            // cout<<head->val<<" "<<fast->val<<endl;
            if(head->val==fast->val){
                head=head->next;
                fast=fast->next;
            // cout<<head->val<<" "<<fast->val<<endl;
            }
            else return 0;
        }
        return 1;
    }
};

File: 234-palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
     ListNode reverseList(ListNode head){
        ListNode dummy=null;

        while(head!=null){
            ListNode  temp=head.next;
          head.next=dummy;
          dummy=head;
          head=temp;
        }

        return dummy;
    }
    public boolean isPalindrome(ListNode head) {
        ListNode slow=head,fast=head;
        ListNode curr=head;
        while(fast!=null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;
        }
        ListNode newhead;
        if(fast==null) newhead=reverseList(slow);
        else newhead=reverseList(slow.next);
        

        while(curr!=null && newhead!=null){
            if(curr.val!=newhead.val)return false;
            curr=curr.next;
            newhead=newhead.next;
        }
        return true;
    }
}